src/main/java/com/google/devtools/build/lib/util/StringEncoding.java

```diff
@@ -0,0 +1,183 @@
+ * different Java types:
+ *
+ * <ul>
+ *   <li>Internal strings: All strings retained by Bazel and used in its inner layers are expected
+ *       to be raw byte sequences stored in Latin-1 {@link String}s. With Java's compact string
+ *       representation, this means that the Latin-1 bytes are stored directly in the internal byte
+ *       array {@link String#value} and the {@link String#coder} is {@link String#LATIN1}.
+ *   <li>Unicode strings: Regular Java strings, which are always Unicode. A common example is a
+ *       {@code string} field in a protobuf message.
+ *   <li>Platform strings: Strings that are passed to or returned from Java (N)IO functions or as
+ *       command-line arguments or environment variables to the {@code java} binary at startup or
+ *       processes started via {@link java.lang.ProcessBuilder}. These strings are encoded and
+ *       decoded by the JVM according to its default native encoding, which is given by the
+ *       {@systemProperty sun.jnu.encoding} system property. With the current JDK version (21), this
+ *       is:
+ *       <ul>
+ *         <li>UTF-8 on macOS;
+ *         <li>determined by the active code page on Windows (Cp1252 on US Windows, can be set to
+ *             UTF-8 by the user);
+ *         <li>determined by the current locale on Linux (forced to en_US.ISO-8859-1 by the client
+ *             if available, otherwise usually UTF-8).
+ *       </ul>
+ *       As a result, there are two cases to consider:
+ *       <ul>
+ *         <li>On Linux with a Latin-1 locale, platform strings are identical to internal strings
+ *             and Java (N)IO functions can be used to operate with Unix API on a raw byte level.
+ *         <li>In all other cases, platform strings are a subset of Unicode strings.
+ *       </ul>
+ * </ul>
+ *
+ * <p>The static methods in this class efficiently reencode {@link String}s between these three
+ * "types". Crucially, since ASCII strings are encoded identically in ISO-8859-1 and UTF-8, such
+ * strings do not need to be reencoded.
+ */
+public final class StringEncoding {
+
+  /**
+   * Transforms an internal string into a platform string as efficiently as possible.
+   *
+   * <p>See the class documentation for more information on the different types of strings.
+   */
+  public static String internalToPlatform(String s) {
+    if (!needsReencodeForPlatform(s)) {
+      return s;
+    }
+    // This is both a performance optimization and a correctness check: internal strings must
+    // always be coded in Latin-1, otherwise they have been constructed out of a non-ASCII string
+    // that hasn't been converted to internal encoding.
+    if (STRING_UNSAFE.getCoder(s) != StringUnsafe.LATIN1) {
+      throw new IllegalArgumentException(
+          "Expected internal string LATIN1 coded, got UTF16 bytes: %s (%s)"
+              .formatted(Arrays.toString(STRING_UNSAFE.getByteArray(s)), s));
```

Comment:
nit: mind imposing an upper limit (dunno, 1000 chars?) on what's printed. Otherwise, it's potentially a lot of log spam.